js/body_mixin.ts

```diff
@@ -0,0 +1,230 @@
+          formdata.append(key, val);
+        }
+      } else {
+        formdata.append(key, String(value));
+      }
+    }
+    return formdata;
+  }
+
+  async text(): Promise<string> {
+    if (typeof this.bodySource === "string") {
+      return this.bodySource;
+    }
+
+    const arr = await this.arrayBuffer();
+    return new TextDecoder("utf-8").decode(arr);
+  }
+
+  async json(): Promise<any> {
+    const raw = await this.text();
+    return JSON.parse(raw);
+  }
+
+  async arrayBuffer(): Promise<ArrayBuffer> {
+    if (
+      this.bodySource instanceof Int8Array ||
+      this.bodySource instanceof Int16Array ||
+      this.bodySource instanceof Int32Array ||
+      this.bodySource instanceof Uint8Array ||
+      this.bodySource instanceof Uint16Array ||
+      this.bodySource instanceof Uint32Array ||
+      this.bodySource instanceof Uint8ClampedArray ||
+      this.bodySource instanceof Float32Array ||
+      this.bodySource instanceof Float64Array
+    ) {
+      return <ArrayBuffer>this.bodySource.buffer;
+    } else if (this.bodySource instanceof ArrayBuffer) {
+      return this.bodySource;
+    } else if (typeof this.bodySource === "string") {
+      const enc = new TextEncoder();
+      return <ArrayBuffer>enc.encode(this.bodySource).buffer;
+    } else if (this.bodySource instanceof ReadableStream) {
+      return bufferFromStream((this.bodySource as ReadableStream).getReader());
+    } else if (this.bodySource instanceof DenoFormData) {
+      const enc = new TextEncoder();
+      return <ArrayBuffer>enc.encode(this.bodySource.toString()).buffer;
+    } else if (!this.bodySource) {
+      return new ArrayBuffer(0);
+    }
+    throw new Error(
+      `Body type not yet implemented: ${this.bodySource.constructor.name}`
+    );
+  }
+}
+
+/** @hidden */
+function validateBodyType(owner: any, bodySource: any) {
+  if (
+    bodySource instanceof Int8Array ||
+    bodySource instanceof Int16Array ||
+    bodySource instanceof Int32Array ||
+    bodySource instanceof Uint8Array ||
+    bodySource instanceof Uint16Array ||
+    bodySource instanceof Uint32Array ||
+    bodySource instanceof Uint8ClampedArray ||
+    bodySource instanceof Float32Array ||
+    bodySource instanceof Float64Array
+  ) {
+    return true;
+  } else if (bodySource instanceof ArrayBuffer) {
+    return true;
+  } else if (typeof bodySource === "string") {
+    return true;
+  } else if (bodySource instanceof ReadableStream) {
+    return true;
+  } else if (bodySource instanceof DenoFormData) {
+    return true;
+  } else if (!bodySource) {
+    return true; // null body is fine
+  }
+  throw new Error(
+    `Bad ${owner.constructor.name} body type: ${bodySource.constructor.name}`
+  );
+}
+
+export function bufferFromStream(
+  stream: ReadableStreamReader
+): Promise<ArrayBuffer> {
+  return new Promise((resolve, reject) => {
+    let parts: Array<Uint8Array> = [];
+    let encoder = new TextEncoder();
+    // recurse
+    (function pump() {
```

Comment:
I am not sure I like this IIFE at all... I need to take a look at the `ReadableStreamReader`, but it feels like this is an async iterator and could be handled much cleaner.  What is the justification for using an IIFE here?